lib/csv/decoder.ex

```diff
@@ -61,120 +61,102 @@ defmodule CSV.Decoder do
       [%{:x => \"a\", :y => \"b\"}, %{:x => \"c\", :y => \"d\"}]
   """
 
-  def decode(stream, options \\ []) do
-    { headers, stream } = options
-                          |> Keyword.get(:headers, false)
-                          |> get_headers(stream, options)
-
-    row_length = case headers do
-      false -> stream |> get_row_length(options)
-      _ -> headers |> Enum.count
-    end
-
-    num_workers = options
-                  |> Keyword.get(:num_workers, options
-                  |> Keyword.get(:num_pipes, Defaults.num_workers))
-
-    multiline_escape = options
-                        |> Keyword.get(:multiline_escape, true)
+  def decode!(stream, options \\ []) do
+    decode(stream, options)
+    |> handle_errors!
+  end
 
+  def decode(stream, options \\ []) do
+    { stream, options |> options_with_defaults }
+    |> prepare_headers
+    |> prepare_row_length
+    |> decode_stream
+  end
 
+  defp decode_stream({ stream, %{ multiline_escape: multiline_escape } = options }) do
     stream
     |> aggregate(multiline_escape)
     |> Stream.with_index
-    |> ParallelStream.map(fn { line, index } ->
-      { line, index }
-      |> Lexer.lex(options)
-      |> lex_to_parse
-      |> Parser.parse(options)
-      |> check_row_length(row_length)
-      |> build_row(headers)
-    end, options |> Keyword.merge(num_workers: num_workers))
-    |> handle_errors
-  end
-  defp aggregate(stream, true) do
-    stream |> LineAggregator.aggregate
-  end
-  defp aggregate(stream, false) do
-    stream
+    |> ParallelStream.map(&(decode_row(&1, options)),
+                          options |> Enum.into([]))
   end
 
-  defp lex_to_parse({ :ok, tokens, index }) do
-    { tokens, index }
-  end
-  defp lex_to_parse(result) do
-    result
+  defp decode_row({ nil, 0 }, %{ row_length: false }), do: { :ok, [] }
+  defp decode_row({ line, index }, %{ headers: headers, row_length: row_length } = options) do
+    with { :ok, parsed, _ } <- parse_row({ line, index }, options |> Enum.into([])),
```

Comment:
Your concern makes sense @beatrichartz.

I do not have Elixir version usage numbers, unfortunately, although my impression is that the community is quick to upgrade. Among the people I know only a few of those with pre-existing big Erlang applications prefer to take some time before upgrading to Erlang 18.

Independently of usage, I'd suggest you seriously consider taking the opportunity of a 2.0 release with a breaking API to also change the Elixir version. This will make 2.x releases painless.

Of course I will go back to `case` if you still prefer to maintain Elixir 1.1 compatibility.